content/terraform/best-practices.md

```diff
@@ -155,8 +241,67 @@ Example:
 
 # Naming Conventions
 
+## Use a programmatically consistent naming convention
+
+All resource names (E.g. things provisioned on AWS) must follow a consistent convention. The reason this is so important is 
+that modules are frequently composed inside of other modules. Enforcing consistency increases the likelihood that modules can 
+invoke other modules without colliding on resource names.
+
+To enforce consistency, we require that all modules use the [`terraform-null-label`](https://github.com/cloudposse/terraform-null-label) module. 
+With this module, users have the ability to change the way resource names are generated such as by changing the order of parameters or the delimiter.
+While the module is opinionated on the parameters, it's proved invaluable as a mechanism for generating consistent resource names.
+
 # DNS Infrastructure
 
+## Use lots of DNS zones
+
+Never mingle DNS from different stages or environments in the same zone.
+
+## Delegate DNS zones across account boundaries
+
+Delegate each AWS account a DNS zone for which it is authoritative. 
+
+## Distinguish between branded domains and service discovery domains
+
+Service discovery domains are what services use to discover each other. These are seldom if ever used by end-users. There should only 
+be one service discovery domain, but there may be many zones delegated from that domain.
+
+Branded domains are the domains that users use to access the services. These are determined by products, marketing, and business use-cases. 
+There may be many branded domains pointing to a single service discovery domain. The architecture of the branded domains won't mirror the 
+service discovery domains.
+
 # Module Design
 
+## Small Opinionated Modules
+
+We believe that modules should do one thing very well. But in order to do that, it requires being opinionated on the design. 
+Simply wrapping terraform resources for the purposes of modularizing code is not that helpful. Implementing a specific use-case
+of those resource is more helpful. 
+
+## Composable Modules
+
+Write all modules to be easily composable into other modules. This is how we're able to achieve economies of scale and stop
+re-inventing the same patterns over and over again.
+
+## Use `variable` inputs
+
+Modules should accept as many parameters as possible. Avoid using inputs of `type = object` has they are harder to document. Of course,
```

Comment:
```suggestion
Modules should accept as many parameters as possible. Avoid using inputs of `type = object` since they are harder to document. Of course,
```